internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -17,32 +31,559 @@ import "context"
+				cp.CallbackUrls = append(cp.CallbackUrls, new.CallbackUrls...)
+			}
+		case "AudClaims":
+			switch {
+			case len(new.AudClaims) == 0:
+				cp.AudClaims = nil
+			default:
+				cp.AudClaims = make([]string, 0, len(new.AudClaims))
+				cp.AudClaims = append(cp.AudClaims, new.AudClaims...)
+			}
+		case "Certificates":
+			switch {
+			case len(new.Certificates) == 0:
+				cp.Certificates = nil
+			default:
+				cp.Certificates = make([]string, 0, len(new.Certificates))
+				cp.Certificates = append(cp.Certificates, new.Certificates...)
+			}
+		}
+	}
+	return cp
+}
+
+// ValidateAuthMethod will test/validate the provided AuthMethod.
 //
 // It will verify that all required fields for a working AuthMethod have values.
 //
-// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, TestAuthMethod
+// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, ValidateAuthMethod
 // retrieves the OpenID Configuration document. The values in the AuthMethod
 // (and associated data) are validated with the retrieved document. The issuer and
 // id token signing algorithm in the configuration are validated with the
-// retrieved document. TestAuthMethod also verifies the authorization, token,
+// retrieved document. ValidateAuthMethod also verifies the authorization, token,
 // and user_info endpoints by connecting to each and uses any certificates in the
 // configuration as trust anchors to confirm connectivity.
 //
 // Options supported are: WithPublicId, WithAuthMethod
-func (r *Repository) TestAuthMethod(ctx context.Context, opt ...Option) error {
-	panic("to-do")
+func (r *Repository) ValidateAuthMethod(ctx context.Context, opt ...Option) error {
+	const op = "oidc.(Repository).TestAuthMethod"
+	opts := getOpts(opt...)
+	var am *AuthMethod
+	switch {
+	case opts.withPublicId != "":
+		var err error
+		am, err = r.lookupAuthMethod(ctx, opts.withPublicId)
+		if err != nil {
+			return errors.Wrap(err, op)
+		}
+	case opts.withAuthMethod != nil:
+		am = opts.withAuthMethod
+	default:
+		return errors.New(errors.InvalidParameter, op, "neither WithPublicId(...) nor WithAuthMethod(...) options were provided")
+	}
+
+	if err := am.isComplete(); err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	// FYI: once converted to an oidc.Provider, any certs configured will be used as trust anchors for all HTTP requests
+	provider, err := convertToProvider(ctx, am)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	info, err := provider.DiscoveryInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	var result *multierror.Error
+	if info.Issuer != am.DiscoveryUrl {
+		result = multierror.Append(result, errors.New(errors.InvalidParameter, op,
+			fmt.Sprintf("auth method issuer doesn't match discovery issuer: expected %s and got %s", am.DiscoveryUrl, info.Issuer)))
+	}
+	for _, a := range am.SigningAlgs {
+		if !strutil.StrListContains(info.IdTokenSigningAlgsSupported, a) {
```

Comment:
The user must enter the correct capitalization for the signing algorithm to be used?

Reply:
Yes.  The spec defines the valid values: https://tools.ietf.org/html/rfc7518#section-3.1